### PR TITLE
[spark] Fix Gloo detecting incorrect Interfaces on DBR

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
When running distributed Pytorch without GPUs, Pytorch selects a **localhost** interface for gloo (i.e. `127.0.0.1:XXX`), breaking distributed training. This method in [Pytorch](https://github.com/pytorch/pytorch/blob/9b88354b80e2b2cce401bd59c7f0b45e13657a0b/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L653) can yield the incorrect interface when a) the the hostname resolves locally to the loopback address or b) when hostname lookups fail.

This is scoped to DBR specifically because `eth0` is guaranteed to exist there. Pytorch+Gloo **does not support** deny-listing like NCCL (as we do in https://github.com/ray-project/ray/pull/31824) because Pytorch directly uses the environment variable `GLOO_SOCKET_IFNAME` as the interface to use https://github.com/pytorch/pytorch/blob/7956ca16e649d86cbf11b6e122090fa05678fac3/torch/csrc/distributed/c10d/init.cpp#L2243. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by WeichenXu123 — COMMENTED** [review]
We need to update here too:

https://github.com/ray-project/ray/blob/f66f1cc0b8c081e243bca126e6810f9317e40f4a/python/ray/util/spark/cluster_init.py#L705

File: python/ray/util/spark/cluster_init.py [hunk 1]
```diff
@@ -703,6 +703,7 @@ def _setup_ray_cluster(
             extra_env={
                 RAY_ON_SPARK_COLLECT_LOG_TO_PATH: collect_log_to_path or "",
                 RAY_ON_SPARK_START_RAY_PARENT_PID: str(os.getpid()),
+                **start_hook.custom_environment_variables(),
             },
         )
```
Comment (ijrsvt) [inline python/ray/util/spark/cluster_init.py:708]: @WeichenXu123 Got it here!

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Comment by WeichenXu123** [comment]
CC @jjyao PR is ready.

**Review by jjyao — APPROVED** [review]

**Pull request merged**